Task 1. Integrate information from the “job_url_data” folder into one dictionary. The keys of the dictionary are individual urls scraped from the website, and the values are the earliest date that the corresponding urls were scraped. Save the results into a json file. How many unique job urls have been collected between May 17, 2022 and May 23, 2022? 

In [192]:
#Import required module
import csv
import json
import os
import re
import pandas as pd
import glob
#Locate the files
directory = 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/'
dictionary = {}
#filter only the files we need
for filename in os.listdir(directory):
    if 'job' in filename:
        csv_read = pd.read_csv(directory + '/' +filename,header=None)
        get_date = fileame.split('_')[4] #get dates from file name and save as list
        for i in range(1,len(csv_read)):
                key = csv_read.iloc[i].values[0]
                if key in dictionary:
                    fetch_date = dictionary[key]
                    if get_date < fetch_date:
                        dictionary[key] = get_date #drop the duplicates
                else:
                    dictionary[key] = get_date
with open("combined_job_url_data.json", "w") as fp:
    json.dump(dictionary , fp) 
len(dictionary) #count the numbers

21260

There are 21260 unique job urls have been collected between May 17, 2022 and May 23, 2022.

Task 2. Clean and integrate information from the “job_info_data” folder into one data frame. Files from this subfolder might have two different formats. Some of them are csv files, while others are json files. The columns might also be named differently. Find ways to read each of the files into pandas, drop records with missing job titles and/or missing job descriptions, and combine them into one dataframe. Lastly, drop records with duplicate job urls, and then save them into a separate csv file. How many unique jobs are there in the cleaned dataframe? 

In [239]:
#check the file's heads in the folder
#locate the files
os.chdir = ('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/')
directory = 'C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/'
#read the json file and drop the nulls
with open ('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5182022_v2.json',encoding='utf—8') as f:
    data182=json.load(f)
df_nested_list18 = pd.json_normalize(data182, record_path =['lnks'])  
df_nested_list18.rename(columns={'link': 'lnks_link', 'job_title': 'lnks_job_title', 'company': 'lnks_company', 'company_url': 'lnks_company_url', 'company_location': 'lnks_company_location', 'job_description': 'lnks_job_description'}, inplace=True)  
cdf_18v2=df_nested_list18.dropna(subset=['lnks_job_title', 'lnks_job_description'])
print(cdf_18v2) #print to check if it's effective

with open ('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5202022_v2.json',encoding='utf—8') as f2:
    data202=json.load(f2)
df_nested_list20 = pd.json_normalize(data202, record_path =['lnks'])  
df_nested_list20.rename(columns={'link': 'lnks_link', 'job_title': 'lnks_job_title', 'company': 'lnks_company', 'company_url': 'lnks_company_url', 'company_location': 'lnks_company_location', 'job_description': 'lnks_job_description'}, inplace=True)  
cdf_20v2=df_nested_list20.dropna(subset=['lnks_job_title', 'lnks_job_description'])
print(cdf_20v2)

#read the csv files
df_17v1 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5172022_v1.csv', low_memory=False)
cdf_17v1=df_17v1.dropna(subset=['lnks_job_title', 'lnks_job_description'])


df_17v2 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5172022_v2.csv', low_memory=False)
cdf_17v2=df_17v2.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_18v1 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5182022_v1.csv', low_memory=False)
cdf_18v1=df_18v1.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_19v1 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5192022_v1.csv', low_memory=False)
cdf_19v1=df_19v1.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_19v2 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5192022_v2.csv', low_memory=False)
cdf_19v2=df_19v2.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_20v1 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5202022_v1.csv', low_memory=False)
cdf_20v1=df_20v1.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_21v1 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5212022_v1.csv', low_memory=False)
cdf_21v1=df_21v1.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_21v2 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5212022_v2.csv', low_memory=False)
cdf_21v2=df_21v2.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_22v1 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5222022_v1.csv', low_memory=False)
cdf_22v1=df_22v1.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_22v2 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5222022_v2.csv', low_memory=False)
cdf_22v2=df_22v2.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_22v3 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5222022_v3.csv', low_memory=False)
cdf_22v3=df_22v3.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_22v4 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5222022_v4.csv', low_memory=False)
cdf_22v4=df_22v4.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_22v5 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5222022_v5.csv', low_memory=False)
cdf_22v5=df_22v5.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_22v6 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5222022_v6.csv', low_memory=False)
cdf_22v6=df_22v6.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_23v1 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5232022_v1.csv', low_memory=False)
cdf_23v1=df_23v1.dropna(subset=['lnks_job_title', 'lnks_job_description'])

df_23v2 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_info_data/parsehub_5232022_v2.csv', low_memory=False)
cdf_23v2=df_23v2.dropna(subset=['lnks_job_title', 'lnks_job_description'])

                                              lnks_link  \
5     https://www.indeed.com/rc/clk?jk=be4f78522cc9e...   
6     https://www.indeed.com/pagead/clk?mo=r&ad=-6NY...   
7     https://www.indeed.com/pagead/clk?mo=r&ad=-6NY...   
8     https://www.indeed.com/pagead/clk?mo=r&ad=-6NY...   
9     https://www.indeed.com/rc/clk?jk=cc5536b8ea743...   
...                                                 ...   
2649  https://www.indeed.com/rc/clk?jk=c065f2f078370...   
2650  https://www.indeed.com/pagead/clk?mo=r&ad=-6NY...   
2651  https://www.indeed.com/rc/clk?jk=9fd411ad9ff45...   
2652  https://www.indeed.com/rc/clk?jk=c0c5ca61a5161...   
2653  https://www.indeed.com/rc/clk?jk=bbfdd84aa83d5...   

                                         lnks_job_title  \
5                    Associate Director Data Management   
6     Safety and Health Specialist - Great Opportuni...   
7                                Sr Material Technician   
8                                1st Assistant Engineer

In [244]:
#combine all the df
job_info_data = pd.concat([cdf_18v2, cdf_20v2,cdf_17v1,cdf_17v2,cdf_18v1,cdf_19v1,cdf_19v2,cdf_20v1,cdf_21v1,cdf_21v2,cdf_22v1,cdf_22v2,cdf_22v3,cdf_22v4,cdf_22v5,cdf_22v6,cdf_23v1,cdf_23v2], axis=0)
cleaned_job_info_data=job_info_data.drop_duplicates(subset=['lnks_link'])
#count the number of unique jobs and save the dataframe into csv file
cleaned_job_info_data.to_csv('cleaned_job_info_data.csv', index=False)
len(cleaned_job_info_data)

16303

In the cleaned dataframe, there are 16303 unique jobs.

Task 3. Merge between “job_url_data” and “job_info_data”. What is the percentage of jobs that can be matched between these two data sources? How are the missing data (unmatched job urls) distributed by date? What about matched job urls? How many complete job listings were we able to collect each day? How would you interpret this result with respect to data quality? Does it mean that our data collection strategy is flawed and thus introduces non-random sampling biases? 

In [354]:
#read requilred files and merge them

files = os.path.join("C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/", "*.csv")
files = glob.glob(files)
print(files)
df = pd.concat(map(pd.read_csv, files), ignore_index=True)
df.drop_duplicates(subset=['job_url'])
print(df)

job_url_data = df.drop_duplicates()

job_url_info= job_url_data.merge(cleaned_job_info_data, how='left', left_on='job_url', right_on='lnks_link')

len(job_url_info)


['C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5172022_v1.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5172022_v2.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5172022_v3.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5172022_v4.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5172022_v5.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5182022_v1.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5192022_v1.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5192022_v2.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5202022_v1.csv', 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data\\job_urls_for_parsehub_5202022_v2.csv', 'C:/Users/27737/compsoc/indee

21260

In [330]:
#count the nulls
job_url_info.isnull().sum()
#the number of null values in lnks_link is 4957
#Count the mach value
result = 21260 - 4957
print(result)

16303


In [332]:
result = 16303/21260
print(result)

0.7668391345249295


The percentage of jobs that can be matched between these two data sources is about 76.68%. 

In [397]:
#How are the missing data (unmatched job urls) distributed by date?
directory = 'C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/'
df1 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5172022_v1.csv', low_memory=False)
df2 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5172022_v2.csv', low_memory=False)
df3 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5172022_v3.csv', low_memory=False)
df4 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5172022_v4.csv', low_memory=False)
df5 = pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5172022_v5.csv', low_memory=False)
  
u517_df = pd.concat([df1, df2,df3,df4,df5])
u517_df['date']= '5172022'
print(u517_df)


u518_df= pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5182022_v1.csv', low_memory=False)
u518_df['date']= '5182022'
print(u518_df)

df7=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5192022_v1.csv', low_memory=False)
df8=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5192022_v2.csv', low_memory=False)
u519_df = pd.concat([df7,df8])
u519_df['date']= '5192022'
print(u519_df)

df9=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5202022_v1.csv', low_memory=False)
df10=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5202022_v2.csv', low_memory=False)
u520_df = pd.concat([df9,df10])
u520_df['date']= '5202022'
print(u520_df)


df11=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5212022_v1.csv', low_memory=False)
df12=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5212022_v2.csv', low_memory=False)
u521_df = pd.concat([df11,df12])
u521_df['date']= '5212022'
print(u521_df)

df13=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5222022_v1.csv', low_memory=False)
df14=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5222022_v2.csv', low_memory=False)
u522_df = pd.concat([df13,df14])
u522_df['date']= '5222022'
print(u522_df)

df15=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5232022_v1.csv', low_memory=False)
df16=pd.read_csv('C:/Users/27737/compsoc/indeed_scraped_data/job_url_data/job_urls_for_parsehub_5232022_v2.csv', low_memory=False)
u523_df = pd.concat([df15,df16])
u523_df['date']= '5232022'
print(u523_df)

url_if= pd.concat([u517_df,u518_df,u519_df,u520_df,u521_df,u522_df,u523_df])
print(url_if)

#drop duplicates 
cleaned_url_df=url_if.drop_duplicates(subset=['date', 'job_url'])
print(cleaned_url_df)

jov_url_df_2=cleaned_url_df.drop_duplicates(subset=['job_url'],keep='first')
print(jov_url_df_2)

                                                job_url     date
0     https://www.indeed.com/pagead/clk?mo=r&ad=-6NY...  5172022
1     https://www.indeed.com/rc/clk?jk=bf3aa4d4d8608...  5172022
2     https://www.indeed.com/rc/clk?jk=ceee36181304b...  5172022
3     https://www.indeed.com/rc/clk?jk=e9735aea38be9...  5172022
4     https://www.indeed.com/company/Swiss-American-...  5172022
...                                                 ...      ...
2266  https://www.indeed.com/pagead/clk?mo=r&ad=-6NY...  5172022
2267  https://www.indeed.com/rc/clk?jk=268837ed32197...  5172022
2268  https://www.indeed.com/pagead/clk?mo=r&ad=-6NY...  5172022
2269  https://www.indeed.com/rc/clk?jk=08d6b1c7f48c1...  5172022
2270  https://www.indeed.com/rc/clk?jk=bd13d1b7b16ed...  5172022

[13177 rows x 2 columns]
                                                job_url     date
0     https://www.indeed.com/rc/clk?jk=a4d8ef765ac38...  5182022
1     https://www.indeed.com/rc/clk?jk=574d165a16772...  5182022

In [407]:
import seaborn as sns
job_url_info_2= jov_url_df_2.merge(cleaned_job_info_data, how='left', left_on='job_url', right_on='lnks_link')

len(job_url_info_2)

job_url_info_2['missing'] = job_url_info_2['lnks_link'].isna()
job_url_info_2

destributed_by_dates = job_url_info_2.groupby(['date', 'missing']).agg({'job_url': 'count'}).reset_index()
destributed_by_dates


date  missing  job_url
0   5172022    False     3905
1   5172022     True     1352
2   5182022    False     1324
3   5182022     True        1
4   5192022    False     3790
5   5192022     True        8
6   5202022    False     4503
7   5202022     True       20
8   5212022    False     2113
9   5212022     True       41
10  5222022    False      668
11  5222022     True      112
12  5232022     True     3423

The missing data is concentrated on May 17th, 22nd and 23rd, and the 23rd has largest number of unmatch value. Meanwhile, most jobs are matched on May 18th,18th and 20th.

In [409]:
#How many complete job listings were we able to collect each day?

clean_data = job_url_info_2.dropna(axis = 0, how ='any')
completed_dates = clean_data.groupby(['date']).agg({'job_url':'count'}).reset_index()
completed_dates

date  job_url
0  5172022     3253
1  5182022     1243
2  5192022     3354
3  5202022     4181
4  5212022     1741
5  5222022      519

In May 17th, we was able to collect 3253 complete job, and there was 1243 complete job we collected on 18th. For 19th, there was 3354, for 20th, there was 4181, for 21st, we had 1741, and for 22nd, there was 519 complete job. Unfortunatly, we failed to collect complete job on 23rd. 

#How would you interpret this result with respect to data quality? 

The percentage of jobs that can be matched is about 76.68% overall, and in 18th and 19th, the number of unmatched job is less than 10. Therefore, even there is a little inconsistent in the 17th and 23rd, the data quality is relatively high.

#Does it mean that our data collection strategy is flawed and thus introduces non-random sampling biases?

Since there are many factor can impact the data quality, if the data collection strategy is consistent, it might not be flawed.And the sampling biases are hard to tell if it is non-random by only going through these data. However, there might be some problems when using these strategies collecting data on 17th and 23rd. Better training of personnel and more focus on coordination and feedback are both likely to improve the quality of data collection, and if there are still days without single matched data, we can say the collecting strategy has problem and try to fix it.